### PR TITLE
encryption: import / export encryption device

### PR DESCRIPTION
**Pull request by miguel-nascimento**
This PR allows the import / export of the encryption device, enables this feature in the sdk and does some changes to use this feature in the stress client.

### CODE REVIEW

File: packages/encryption/src/encryptionDevice.ts [hunk 1]
```diff
@@ -135,6 +157,34 @@ export class EncryptionDevice {
         )
     }
 
+    private async initializeFromExportedDevice(
```
Comment (miguel-nascimento) [inline packages/encryption/src/encryptionDevice.ts:160]: rewrite of [matrix implementation](https://github.com/matrix-org/matrix-js-sdk/blob/0fe53876ec555482bbeb086fd4dec8de7b2e21eb/src/crypto/OlmDevice.ts#L224), I wonder if the outbound/inbound sessions import here are the same of #257

Comment (miguel-nascimento) [reply packages/encryption/src/encryptionDevice.ts:160]: if thats true, this PR alone should close #603

File: packages/stress/src/utils/stressClient.ts [hunk 2]
```diff
@@ -235,12 +254,22 @@ export class StressClient {
...
 
     async stop() {
+        await this.exportDevice()
         await this.streamsClient.stop()
     }
+
+    async exportDevice(): Promise<ExportedDevice | undefined> {
+        const device = await this.streamsClient.cryptoBackend?.encryptionDevice.exportDevice()
+        if (device) {
+            await fs.writeFile(this.deviceFilePath, JSON.stringify(device, null, 2))
```
Comment (texuf) [inline packages/stress/src/utils/stressClient.ts:270]: can you try catch around the write file?

File: packages/stress/src/utils/stressClient.ts [hunk 3]
```diff
@@ -186,12 +194,23 @@ export class StressClient {
         return channelId
     }
 
-    async startStreamsClient(metadata: { spaceId: string }) {
+    async startStreamsClient(config: Parameters<typeof this.streamsClient.initializeUser>[0]) {
```
Comment (texuf) [inline packages/stress/src/utils/stressClient.ts:197]: can you just make a normal parameter list here?

File: packages/sdk/src/client.ts [hunk 4]
```diff
@@ -316,7 +320,12 @@ export class Client
         this.syncedStreamsExtensions.setStreamIds(streamIds)
     }
 
-    async initializeUser(newUserMetadata?: { spaceId: Uint8Array | string }): Promise<void> {
+    async initializeUser(
+        opts?: {
+            metadata?: { spaceId: Uint8Array | string }
+        } & EncryptionDeviceInitOpts,
+    ): Promise<void> {
```
Comment (texuf) [inline packages/sdk/src/client.ts:327]: just change `newUserMetadata?: { spaceId: Uint8Array | string }`
to `opts?: { spaceId: Uint8Array | string, encryptionDevice?: EncryptionDeviceInitOpts }`

Comment (texuf) [reply packages/sdk/src/client.ts:327]: and delete line 331

File: packages/stress/src/utils/stressClient.ts [hunk 5]
```diff
@@ -186,12 +194,23 @@ export class StressClient {
...
         if (isDefined(this.streamsClient.userStreamId)) {
             return
         }
-        await this.streamsClient.initializeUser(metadata)
+        let device: ExportedDevice | undefined
+        try {
+            const rawDevice = await fs.readFile(this.deviceFilePath, 'utf8')
+            device = JSON.parse(rawDevice) as ExportedDevice
+        } catch {
+            return
```
Comment (texuf) [inline packages/stress/src/utils/stressClient.ts:206]: You ignored the error here! :)

File: packages/stress/src/utils/stressClient.ts (hunk 2 above)
Comment (miguel-nascimento) [reply packages/stress/src/utils/stressClient.ts]: nice catch! *ba dum tss

File: packages/stress/src/utils/stressClient.ts (hunk 5 above)
Comment (miguel-nascimento) [reply packages/stress/src/utils/stressClient.ts]: thank you!! 🙏

**Review by texuf — APPROVED** [review]

**Pull request merged**